Reference type of 'function definition' is defined as 'FbasicType', but it must be 'FfunctionType'

The following code is parsed correctly by `F_Front` but throw an exception at decompile time. 

```fortran 
MODULE mod1
CONTAINS

  SUBROUTINE sub1(param1)
    REAL(KIND=8), INTENT(OUT) :: param1
  
    param1 = 8.0
  END SUBROUTINE sub1

  FUNCTION param1()
    REAL (KIND=8) :: param1
 
    param1  =  8.0
  END FUNCTION param1
END MODULE mod1
```

Problem comes from the `type` in the name node. It should refer to the `FfunctionType(F7f9932409f40)` and not the `FbaiscType`
```xml
<FfunctionType type="F7f9932409f40" return_type="R7f993240a140">
  <params>
  </params>
</FfunctionType>
<!-- omitted xml here -->
<FfunctionDefinition lineno="10" file="dummy.f90">
   <name type="R7f9932408ef0">param1</name>
   <symbols>
     <id type="F7f9932409f40" sclass="ffunc">
       <name>param1</name>
     </id>
   </symbols>
   <declarations>
     <varDecl lineno="10" file="dummy.f90">
       <name type="F7f9932409f40">param1</name>
     </varDecl>
   </declarations>
  <body>
    <FassignStatement lineno="13" file="dummy.f90">
      <Var type="R7f993240a140" scope="local">param1</Var>
      <FrealConstant type="Freal">8.0</FrealConstant>
    </FassignStatement>
  </body>
</FfunctionDefinition>
```

```bash
Exception in thread "main" xcodeml.util.XmException: xcodeml.util.XmTranslationException: <FfunctionDefinition> Error(4): Reference type of 'function definition' is defined as 'FbasicType', but it must be 'FfunctionType'.
<Element Tree>
#document
  +XcodeProgram
    +globalDeclarations
      +FmoduleDefinition
        +FcontainsStatement
          +FfunctionDefinition

	at xcodeml.f.decompile.XfDecompileDomVisitor.invokeEnter(XfDecompileDomVisitor.java:104)
	at xcodeml.f.util.XmfDecompiler.decompile(XmfDecompiler.java:79)
	at xcodeml.util.XmBackEnd.run(XmBackEnd.java:215)
	at xcodeml.f.util.omx2f.main(omx2f.java:19)
Caused by: xcodeml.util.XmTranslationException: <FfunctionDefinition> Error(4): Reference type of 'function definition' is defined as 'FbasicType', but it must be 'FfunctionType'.
<Element Tree>
#document
  +XcodeProgram
    +globalDeclarations
      +FmoduleDefinition
        +FcontainsStatement
          +FfunctionDefinition

	at xcodeml.f.decompile.XfDecompileDomVisitor.fail(XfDecompileDomVisitor.java:1374)
	at xcodeml.f.decompile.XfDecompileDomVisitor.findFunctionType(XfDecompileDomVisitor.java:3419)
	at xcodeml.f.decompile.XfDecompileDomVisitor$FfunctionDefinitionVisitor.enter(XfDecompileDomVisitor.java:3470)
	at xcodeml.f.decompile.XfDecompileDomVisitor._enter(XfDecompileDomVisitor.java:1368)
	at xcodeml.f.decompile.XfDecompileDomVisitor.invokeEnter(XfDecompileDomVisitor.java:826)
	at xcodeml.f.decompile.XfDecompileDomVisitor._invokeEnter(XfDecompileDomVisitor.java:714)
	at xcodeml.f.decompile.XfDecompileDomVisitor.access$1600(XfDecompileDomVisitor.java:31)
	at xcodeml.f.decompile.XfDecompileDomVisitor$FcontainsStatementVisitor.enter(XfDecompileDomVisitor.java:2626)
	at xcodeml.f.decompile.XfDecompileDomVisitor._enter(XfDecompileDomVisitor.java:1368)
	at xcodeml.f.decompile.XfDecompileDomVisitor.invokeEnter(XfDecompileDomVisitor.java:826)
	at xcodeml.f.decompile.XfDecompileDomVisitor.access$500(XfDecompileDomVisitor.java:31)
	at xcodeml.f.decompile.XfDecompileDomVisitor$FmoduleDefinitionVisitor.enter(XfDecompileDomVisitor.java:4102)
	at xcodeml.f.decompile.XfDecompileDomVisitor._enter(XfDecompileDomVisitor.java:1368)
	at xcodeml.f.decompile.XfDecompileDomVisitor.invokeEnter(XfDecompileDomVisitor.java:826)
	at xcodeml.f.decompile.XfDecompileDomVisitor._invokeChildEnter(XfDecompileDomVisitor.java:810)
	at xcodeml.f.decompile.XfDecompileDomVisitor.access$100(XfDecompileDomVisitor.java:31)
	at xcodeml.f.decompile.XfDecompileDomVisitor$GlobalDeclarationsVisitor.enter(XfDecompileDomVisitor.java:1786)
	at xcodeml.f.decompile.XfDecompileDomVisitor._enter(XfDecompileDomVisitor.java:1368)
	at xcodeml.f.decompile.XfDecompileDomVisitor.invokeEnter(XfDecompileDomVisitor.java:826)
	at xcodeml.f.decompile.XfDecompileDomVisitor._invokeChildEnter(XfDecompileDomVisitor.java:810)
	at xcodeml.f.decompile.XfDecompileDomVisitor.access$100(XfDecompileDomVisitor.java:31)
	at xcodeml.f.decompile.XfDecompileDomVisitor$XcodeProgramVisitor.enter(XfDecompileDomVisitor.java:1416)
	at xcodeml.f.decompile.XfDecompileDomVisitor._enter(XfDecompileDomVisitor.java:1368)
	at xcodeml.f.decompile.XfDecompileDomVisitor.invokeEnter(XfDecompileDomVisitor.java:102)
	... 3 more
Caused by: java.lang.Exception
	at xcodeml.f.decompile.XmfDecompilerContext.setLastErrorMessage(XmfDecompilerContext.java:141)
	at xcodeml.f.decompile.XfDecompileDomVisitor.findFunctionType(XfDecompileDomVisitor.java:3413)
	... 25 more
```

--- a/F-FrontEnd/src/F-compile-decl.c
+++ b/F-FrontEnd/src/F-compile-decl.c
@@ -1314,7 +1314,14 @@ declare_external_proc_id(SYMBOL s, TYPE_DESC tp, int def_flag)
     assert(ep != NULL);
 
     if (def_flag == TRUE || EXT_PROC_TYPE(ep) == NULL) {
-        EXT_PROC_TYPE(ep) = tp;
+        if(tp != NULL 
+            && (TYPE_BASIC_TYPE(tp) == TYPE_FUNCTION 
+            || TYPE_BASIC_TYPE(tp) == TYPE_SUBR 
+            || TYPE_BASIC_TYPE(tp) ==  TYPE_MODULE))
+        {
+            // Do not set non-function type to PROC
+            EXT_PROC_TYPE(ep) = tp;
+        }
     } else if(EXT_IS_DEFINED(ep)) {
         /* avoid overwriting EXT_ID already defined. */
         ep = new_external_id_for_external_decl(s, tp);

--- a/F-FrontEnd/test/testdata/issue279.f90
+++ b/F-FrontEnd/test/testdata/issue279.f90
@@ -1,0 +1,15 @@
+MODULE mod1
+CONTAINS
+
+  SUBROUTINE sub1(param1)
+    REAL(KIND=8), INTENT(OUT) :: param1
+  
+    param1 = 8.0
+  END SUBROUTINE sub1
+
+  FUNCTION param1()
+    REAL (KIND=8) :: param1
+ 
+    param1  =  8.0
+  END FUNCTION param1
+END MODULE mod1

